runtime/src/bank.rs

```diff
@@ -360,16 +360,16 @@ impl Bank {
         self.epoch_schedule.get_epoch(self.slot)
     }
 
-    pub fn freeze_lock(&self) -> RwLockReadGuard<Hash> {
+    pub fn freeze_lock(&self) -> RwLockReadGuard<(Hash, bool)> {
         self.hash.read().unwrap()
     }
 
     pub fn hash(&self) -> Hash {
-        *self.hash.read().unwrap()
+        self.hash.read().unwrap().0
     }
 
     pub fn is_frozen(&self) -> bool {
-        *self.hash.read().unwrap() != Hash::default()
+        self.hash.read().unwrap().1
```

Comment:
was Hash::default() expensive?

Reply:
This specific comparison is not, but I want to have a modified state after `bank.freeze()` and since I am not computing `bank::hash` directly in freeze, I use this bool to track if the bank is frozen and the hash will be updated at a later time.

Reply:
I made the logic a bit clearer, maybe it's more obvious now.

Reply:
looks like votes for this frozen bank (votes going into one of its children) will fail until we finish with hashing this bank, and the timing of hashing will differ from node to node...

Reply:
Yes, the vote will not be submitted until the hash operation finishes. Timing of hashing differs from node to node even before this change.

Reply:
I see. We could queue the votes until the hash operation is complete.

Reply:
> Yes, the vote will not be submitted until the hash operation finishes. Timing of hashing differs from node to node even before this change.

"timing of hashing" relative to when the chain continues is the same from node to node without this change.  there can be no child banks of a bank until it finishes hashing and updates slot_hashes.  slot_hashes are "on chain", in that they are presented in an account to on-chain programs.  they need to be the same across all nodes at every bank height.

the only way to guarantee this will be correct is to hold any bank processing until slot_hashes is updated, because *any* transaction can load that account and needs to see the same values

Reply:
Well it can hold any transactions that load that account.

Reply:
it has to hold the entire Bank if we are to preserve transaction batching...